identity-iota/src/tangle/explorer.rs

```diff
@@ -0,0 +1,240 @@
+}
+
+/// A Tangle explorer URL.
+#[derive(Clone, Debug, PartialEq, Eq, Deserialize, Serialize)]
+#[repr(transparent)]
+#[serde(transparent)]
+pub struct ExplorerUrl(Url);
+
+/// A Tangle explorer URL with convenience functions for constructing URLs for viewing
+/// published messages or IOTA DIDs.
+///
+/// # Example
+///
+/// ```
+/// # use identity_iota::did::IotaDID;
+/// # use identity_iota::tangle::ExplorerUrl;
+/// let explorer = ExplorerUrl::mainnet();
+/// let did = IotaDID::parse("did:iota:H3C2AVvLMv6gmMNam3uVAjZpfkcJCwDwnZn6z3wXmqPV").unwrap();
+/// assert_eq!(
+///   explorer.resolver_url(&did).unwrap(),
+///   "https://explorer.iota.org/mainnet/identity-resolver/did:iota:H3C2AVvLMv6gmMNam3uVAjZpfkcJCwDwnZn6z3wXmqPV"
+/// );
+/// ```
+impl ExplorerUrl {
+  /// Constructs a new Tangle explorer URL.
+  ///
+  /// Use [`ExplorerUrl::mainnet`] or [`ExplorerUrl::devnet`] unless using a private Tangle
+  /// or local explorer.
+  ///
+  /// NOTE: does not validate whether this corresponds to an actual Tangle explorer.
+  pub fn new(url: Url) -> Result<Self> {
+    if url.cannot_be_a_base() {
+      return Err(Error::InvalidExplorerURL);
+    }
+    Ok(Self(url))
+  }
+
+  /// Constructs a new Tangle explorer URL from a string.
+  ///
+  /// See [`ExplorerUrl::new`].
+  ///
+  /// # Example
+  ///
+  /// Point to a Tangle explorer deployed locally.
+  /// ```
+  /// # use identity_iota::tangle::ExplorerUrl;
+  /// let explorer = ExplorerUrl::parse("http://127.0.0.1:8082/").unwrap();
+  /// ```
+  pub fn parse(url: &str) -> Result<Self> {
+    let url: Url = Url::parse(url).map_err(|_| Error::InvalidExplorerURL)?;
+    Self::new(url)
+  }
+
+  /// Returns the Tangle explorer URL for the mainnet.
+  #[inline]
+  pub fn mainnet() -> &'static ExplorerUrl {
+    &EXPLORER_MAIN
+  }
+
+  /// Returns the Tangle explorer URL for the devnet.
+  #[inline]
+  pub fn devnet() -> &'static ExplorerUrl {
+    &EXPLORER_DEV
+  }
+
+  /// Returns the web explorer URL of the given `message_id`.
+  ///
+  /// E.g. https://explorer.iota.org/mainnet/message/<message_id>
+  pub fn message_url(&self, message_id: &MessageId) -> Result<Url> {
+    let mut url: Url = self.0.clone();
+    url
+      .path_segments_mut()
+      .map_err(|_| Error::InvalidExplorerURL)?
+      .push("message")
+      .push(&message_id.to_string());
+    Ok(url)
+  }
+
+  /// Returns the web identity resolver URL for the given DID.
+  ///
+  /// E.g. https://explorer.iota.org/mainnet/identity-resolver/<did>
+  pub fn resolver_url(&self, did: &impl DID) -> Result<Url> {
+    let mut url: Url = self.0.clone();
+    url
+      .path_segments_mut()
+      .map_err(|_| Error::InvalidExplorerURL)?
+      .push("identity-resolver")
+      .push(did.as_str());
+    Ok(url)
+  }
+}
+
+impl AsRef<str> for ExplorerUrl {
+  fn as_ref(&self) -> &str {
+    self.0.as_ref()
+  }
+}
+
+impl Deref for ExplorerUrl {
+  type Target = Url;
+
+  fn deref(&self) -> &Self::Target {
+    &self.0
+  }
+}
```

Comment:
Convenience for using it as a regular `Url`, similar to how the crate `Url` implements `Deref` for `::url::Url`. This also only implements `Deref` not `DerefMut` so there's little chance of breaking any invariants (even though the only one we enforce is that it is not a base url, and even that is just to avoid a runtime error with `path_segment_mut`, so it's still enforced).

I'll remove it though since I know there's contention around `Deref`.

Reply:
I'm fine with `Deref` per se, but I'm not sure whether it's truly needed here. The type is probably unlikely to change, but if it did, the absence of `Deref` would make it much easier to change internals.